robot/impl/local_robot_test.go

```diff
@@ -2389,3 +2392,264 @@ func TestOrphanedResources(t *testing.T) {
+			{
+				Name:    "mod",
+				ExePath: rutils.ResolveFile("examples/customresources/demos/complexmodule/run.sh"),
+			},
+		},
+		Components: []config.Component{
+			{
+				Name:      "g",
+				Namespace: "acme",
+				Model:     gizmoModel,
+				Type:      resource.SubtypeName("gizmo"),
+				DependsOn: []string{"m"},
+			},
+			{
+				Name:  "m",
+				Model: fakeModel,
+				Type:  motor.SubtypeName,
+			},
+			{
+				Name:      "m1",
+				Model:     fakeModel,
+				Type:      motor.SubtypeName,
+				DependsOn: []string{"g"},
+			},
+		},
+	}
+	r, err := robotimpl.New(ctx, cfg, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, r.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	// Assert that reconfiguring module 'mod' to a new module that does not handle
+	// 'g' removes modular component 'g' and its dependent 'm1' and leaves 'm' as-is.
+	cfg2 := &config.Config{
+		Modules: []config.Module{
+			{
+				Name:    "mod",
+				ExePath: rutils.ResolveFile("examples/customresources/demos/simplemodule/run.sh"),
+			},
+		},
+		Components: []config.Component{
+			{
+				Name:      "g",
+				Namespace: "acme",
+				Model:     gizmoModel,
+				Type:      resource.SubtypeName("gizmo"),
+				DependsOn: []string{"m"},
+			},
+			{
+				Name:  "m",
+				Model: fakeModel,
+				Type:  motor.SubtypeName,
+			},
+			{
+				Name:      "m1",
+				Model:     fakeModel,
+				Type:      motor.SubtypeName,
+				DependsOn: []string{"g"},
+			},
+		},
+	}
+	r.Reconfigure(ctx, cfg2)
+
+	res, err := r.ResourceByName(gizmoapi.Named("g"))
+	test.That(t, err, test.ShouldBeError,
+		rutils.NewResourceNotFoundError(gizmoapi.Named("g")))
+	test.That(t, res, test.ShouldBeNil)
+	res, err = r.ResourceByName(motor.Named("m1"))
+	test.That(t, err, test.ShouldBeError,
+		rutils.NewResourceNotFoundError(motor.Named("m1")))
+	test.That(t, res, test.ShouldBeNil)
+	_, err = r.ResourceByName(motor.Named("m"))
+	test.That(t, err, test.ShouldBeNil)
+
+	// Remove module entirely.
+	cfg3 := &config.Config{
+		Components: []config.Component{
+			{
+				Name:      "g",
+				Namespace: "acme",
+				Model:     gizmoModel,
+				Type:      resource.SubtypeName("gizmo"),
+				DependsOn: []string{"m"},
+			},
+			{
+				Name:  "m",
+				Model: fakeModel,
+				Type:  motor.SubtypeName,
+			},
+			{
+				Name:      "m1",
+				Model:     fakeModel,
+				Type:      motor.SubtypeName,
+				DependsOn: []string{"g"},
+			},
+		},
+	}
+	r.Reconfigure(ctx, cfg3)
+
+	// Assert that adding module 'mod' back with original executable path re-adds
+	// modular component 'g' and its dependent 'm1', and that 'm' is still present.
+	r.Reconfigure(ctx, cfg)
+
+	_, err = r.ResourceByName(gizmoapi.Named("g"))
+	test.That(t, err, test.ShouldBeNil)
+	_, err = r.ResourceByName(motor.Named("m1"))
+	test.That(t, err, test.ShouldBeNil)
+	m, err := r.ResourceByName(motor.Named("m"))
+	test.That(t, err, test.ShouldBeNil)
+
+	// Assert that motor 'm' can still be used.
```

Comment:
I think we actually want to test `m1` here. I'd like to see a more substantive test - and you may need to create a custom implementation of something that depends on the `g` so that we can see that the child resource is behaving correctly (a generic resource that calls `g` and returns the result is good enough)

Reply:
Sounds good. Replaced `m1` with a `doodad` that depends on `g` and has a method `doThroughGizmo` that can make gRPC calls through `g`; everything seems to work.